src/engine/enginemaster.cpp

```diff
@@ -443,48 +530,212 @@ void EngineMaster::process(const int iBufferSize) {
+        // (headphones, master, booth, and record/broadcast), delay the outputs
+        // by the input latency before mixing the inputs into them.
+
+        // This way, if the microphone user is on beat with the output they hear from
+        // the speakers/headphones, the input will remain on beat at the expense of
+        // adding some latency to Mixxx's output (no latency will be added to the
+        // microphone signal though). If no microphone inputs are configured, avoid adding
+        // this extra latency.
+
+        // If the input signal is being directly monitored (mixed with the master output
+        // in hardware), do not delay the audible outputs. However, delay the
+        // record/broadcast signal to keep it on beat and consistent with what
+        // the user hears from the speakers/headphones.
+        if (configuredTalkoverMixMode == TalkoverMixMode::MASTER) {
+            // Apply input latency compensation to master mix
+            // Do not add unnecessary latency if no microphones are configured
+            if (m_pNumMicsConfigured->get() > 0) {
+                m_pInputLatencyCompensationDelay->process(m_pMaster, iBufferSize);
+            }
+
+            // Copy master mix to booth output with booth gain before mixing
+            // talkover with master mix
+            if (boothEnabled) {
+                CSAMPLE boothGain = m_pBoothGain->get();
+                if (m_bRampingGain) {
+                    SampleUtil::copy1WithRampingGain(m_pBooth, m_pMaster,
+                        m_boothGainOld, boothGain, iBufferSize);
+                } else {
+                    SampleUtil::copy1WithGain(m_pBooth, m_pMaster,
+                        boothGain, iBufferSize);
+                }
+                m_boothGainOld = boothGain;
+            }
+
+            // Mix talkover into master mix
+            if (m_pNumMicsConfigured->get() > 0) {
+                SampleUtil::copy2WithGain(m_pMaster,
+                    m_pMaster, 1.0,
                     m_pTalkover, 1.0,
                     iBufferSize);
-        } else {
-            SampleUtil::copy3WithGain(m_pMaster,
-                    m_pOutputBusBuffers[EngineChannel::LEFT], 1.0,
-                    m_pOutputBusBuffers[EngineChannel::CENTER], 1.0,
-                    m_pOutputBusBuffers[EngineChannel::RIGHT], 1.0,
+            }
+
+            // Process master channel effects
+            if (m_pEngineEffectsManager) {
+                GroupFeatureState masterFeatures;
+                // Well, this is delayed by one buffer (it's dependent on the
+                // output). Oh well.
+                if (m_pVumeter != NULL) {
+                    m_pVumeter->collectFeatures(&masterFeatures);
+                }
+                masterFeatures.has_gain = true;
+                masterFeatures.gain = m_pMasterGain->get();
+                m_pEngineEffectsManager->process(m_masterHandle.handle(), m_pMaster,
```

Comment:
Master effects must be processed after mixing in talkover. Applying master effects to the booth could be done, but it would require double processing the master effects, which would require merging #1254 first.